### ISSUE TITLE
Error compiling iOS tests with 5.0.0.M2 and compiler plugin

### ISSUE THREAD

**Issue by drampelt** [kotest/kotest#2565]
**Which version of Kotest are you using**
5.0.0.M2

I'm trying out the 5.0.0.M2 preview version with the gradle and compiler plugin, it works for JS and JVM but I've having trouble getting it working for iOS. Trying to run the `iosX64Test` task results in the following error during the linking step:

```
The /Applications/Xcode-12.5.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/bin/ld command returned non-zero exit code: 1.
output:
Undefined symbols for architecture x86_64:
  "_kfun:io.kotest.core.spec.DisplayName#<get-name>(){}kotlin.String", referenced from:
      _kfun:io.kotest.engine.test.names.DefaultDisplayNameFormatter#format(kotlin.reflect.KClass<*>){}kotlin.String in result.o
ld: symbol(s) not found for architecture x86_64
```

Am I missing a depedency specific to iOS? I have the `kotest-framework-engine` dependency in my `commonTest` source set.

**Comment by sksamuel** [comment]
I would check you're using 5.0.0.m2 for everything

**Comment by sksamuel** [comment]
Ok I get the same on another project.

e: /home/sam/.konan/dependencies/x86_64-unknown-linux-gnu-gcc-8.3.0-glibc-2.19-kernel-4.9-2/x86_64-unknown-linux-gnu/bin/ld.gold invocation reported errors

The /home/sam/.konan/dependencies/x86_64-unknown-linux-gnu-gcc-8.3.0-glibc-2.19-kernel-4.9-2/x86_64-unknown-linux-gnu/bin/ld.gold command returned non-zero exit code: 1.
output:
/home/runner/work/kotest/kotest/kotest-framework/kotest-framework-engine/src/commonMain/kotlin/io/kotest/engine/test/names/DefaultDisplayNameFormatter.kt:76: error: undefined reference to 'kfun:io.kotest.core.spec.DisplayName#<get-name>(){}kotlin.String'


It looks like the way Kotlin encodes annotations has changed. I'll see if 1.6 fixes this but that's blocked on another bug atm.

**Comment by charleskorn** [comment]
I am also seeing this issue with M2 and M3, but had no issues on M1. I'm seeing this on Windows, Mac and Linux targets.

I'm also getting a similar error message for another symbol:

```
The /home/runner/.konan/dependencies/x86_64-unknown-linux-gnu-gcc-8.3.0-glibc-2.19-kernel-4.9-2/x86_64-unknown-linux-gnu/bin/ld.gold command returned non-zero exit code: 1.
output:
/home/runner/work/kotest/kotest/kotest-framework/kotest-framework-engine/src/commonMain/kotlin/io/kotest/engine/spec/interceptor/ApplyExtensionsInterceptor.kt:29: error: undefined reference to 'kfun:io.kotest.core.extensions.ApplyExtension#<get-extensions>(){}kotlin.Array<out|kotlin.reflect.KClass<out|io.kotest.core.extensions.Extension>>'
/home/runner/work/kotest/kotest/kotest-framework/kotest-framework-engine/src/commonMain/kotlin/io/kotest/engine/test/names/DefaultDisplayNameFormatter.kt:76: error: undefined reference to 'kfun:io.kotest.core.spec.DisplayName#<get-name>(){}kotlin.String'
```

**Comment by sksamuel** [comment]
I'm going to cut another milestone release but in the meantime, the latest snapshots are built with a 1.6 rc2 preview.

**Comment by sksamuel** [comment]
Is 5.0.0.M4 working ?

**Comment by charleskorn** [comment]
> Is 5.0.0.M4 working ?

Not for me - with 5.0.0.M4 and the 1.6 release version of Kotlin, I'm still getting [errors](https://github.com/batect/docker-client/runs/4209492025?check_suite_focus=true#step:13:325):

```
The /home/runner/.konan/dependencies/x86_64-unknown-linux-gnu-gcc-8.3.0-glibc-2.19-kernel-4.9-2/x86_64-unknown-linux-gnu/bin/ld.gold command returned non-zero exit code: 1.
output:
/home/runner/work/kotest/kotest/kotest-framework/kotest-framework-engine/src/commonMain/kotlin/io/kotest/engine/spec/interceptor/ApplyExtensionsInterceptor.kt:28: error: undefined reference to 'kfun:io.kotest.core.extensions.ApplyExtension#<get-extensions>(){}kotlin.Array<out|kotlin.reflect.KClass<out|io.kotest.core.extensions.Extension>>'
/home/runner/work/kotest/kotest/kotest-framework/kotest-framework-engine/src/commonMain/kotlin/io/kotest/engine/spec/interceptor/EnabledIfSpecInterceptor.kt:34: error: undefined reference to 'kfun:io.kotest.core.annotation.EnabledIf#<get-enabledIf>(){}kotlin.reflect.KClass<out|io.kotest.core.annotation.EnabledCondition>'
/mnt/agent/work/6326934d18cfe24e/kotlin/libraries/stdlib/common/src/generated/_Arrays.kt:46: error: undefined reference to 'kfun:io.kotest.core.annotation.RequiresTag#<get-values>(){}kotlin.Array<out|kotlin.String>'
/home/runner/work/kotest/kotest/kotest-framework/kotest-framework-engine/src/commonMain/kotlin/io/kotest/engine/test/names/DefaultDisplayNameFormatter.kt:75: error: undefined reference to 'kfun:io.kotest.core.spec.DisplayName#<get-name>(){}kotlin.String'
```

**Comment by sksamuel** [comment]
Can you try 5.0.0.RC.
I've made a change that hopefully works around this compiler bug.

**Comment by charleskorn** [comment]
Still no luck with 5.0.0.RC. I've also updated to version 5.0.0.6 of the Gradle plugin, but that didn't help either.

Sample build logs are available [here](https://github.com/batect/docker-client/runs/4243255827?check_suite_focus=true#step:13:324).

**Comment by sksamuel** [comment]
You are on 1.6 final ?

**Comment by charleskorn** [comment]
Yep, [this](https://github.com/batect/docker-client/blob/renovate/kotlin/client/build.gradle.kts) is my Gradle project if you want to take a look.

**Comment by sksamuel** [comment]
I would say its a compiler issue, because nothing in kotest should be
causing the compiler to fail on annotations. I added in a work around to
avoid those annotations if not on the JVM but I guess that doesn't help.

Let's file a bug with jetbrains.

**Comment by charleskorn** [comment]
Sure, I'll pull together a sample project and file a bug with them.

**Comment by sksamuel** [comment]
Appreciate it. You can tag me in it too if you want @sksamuel.

**Comment by charleskorn** [comment]
I wasn't able to tag you, but I've raised this as https://youtrack.jetbrains.com/issue/KT-49790.

**Comment by sksamuel** [comment]
Thanks Charles

**Comment by sksamuel** [comment]
Looks like they have a minimal reproducer, so hopefully fixed in 1.6.10

**Comment by sbogolepov** [comment]
Hi. First of all, thanks for the report :)
Indeed, we reduced the problem to a minimal reproducer. But it does not mean that this is going to be fixed in 1.6.10. The purpose of 1.6.10 is to fix regressions that were introduced in 1.6.0, and this bug was introduced way-way earlier. It is definitely going to be fixed in 1.6.20, though (in fact, it should be already fixed in compiler dev builds).

The root cause is that Kotlin/Native does not generate property accessors for annotation classes. Is it possible to workaround this bug on kotest's side until 1.6.20 is released?

**Comment by sksamuel** [comment]
Any suggestions on a workaround? I tried avoiding the paths that use the
annotations unless the platform is JVM.

**Comment by sbogolepov** [comment]
Well, this is a compile-time error, so even if these paths are never executed, they still will be compiled and trigger the error.
The idea is to make sure that calls to property accessors are not compiled on Native. For example, you can create expect/actual wrappers around annotation properties:
`commonMain`:
```kotlin
expect val RequiresTag.valuesWrapper: Array<out String>
```
`jsMain` and `jvmMain`:
```kotlin
actual val RequiresTag.valuesWrapper: Array<out String>
    get() = values
```
`desktopMain`:
```kotlin
actual val RequiresTag.valuesWrapper: Array<out String>
    get() = throw UnsupportedOperationException("Annotation properties should not be called in Kotlin/Native!")
```
And use `valuesWrapper` instead of `values` in the common code.
I'm not familiar with kotest sources, so my suggestion is probably not the prettiest one. Sorry for that :)

**Comment by sksamuel** [comment]
So the annotations stay as they are, but we only access via those extension vals.
I can give it a go.

**Comment by sksamuel** [comment]
Seems you can only have one wrapper per package as well (FYI)

**Comment by sksamuel** [comment]
@charleskorn can you try build 5.0.0.794-SNAPSHOT ?
if that works I will publish final RC before 5.0.0

**Comment by charleskorn** [comment]
Good news: that now compiles.

However, running the tests produces no progress information during the run from Gradle, and no tests in the test report. Failing tests also don't cause the task to fail. There's also none of the usual `##teamcity` output from kotest if I run the `.kexe` directly. (The same tests fail as expected on the JVM.)

**Comment by sksamuel** [comment]
Yeah I'm seeing something similar myself.
I'll have to binary search to find out what's gone on

**Comment by sksamuel** [comment]
Ok its fixed locally. Try 795 once it completes building and you should be good for native. I can run native tests with 1.6, those annotation errors have gone away (I use linux btw). Let me know :)

There's still an issue with JS that has started with 1.6. Some runtime issue with coroutines. If you're not using JS then it won't affect you, but I hold off releasing 5.0.0 final until I lock down what it is.

**Comment by sksamuel** [comment]
Ok the issue with JS is only when using node. So I guess that means its probably in the 1.6 compiler somewhere.
I've raised this: https://youtrack.jetbrains.com/issue/KT-49851

**Comment by charleskorn** [comment]
I'm now getting tests running with that version 🎉 However it looks like I'm also getting tests running multiple times in the wrong class. It seems to be additive - tests from the first test class that runs seem to then run again in the context of the second class, then tests from the first two classes run again in the context of the third and so on.

**Comment by sksamuel** [comment]
I see the same, next build should have this wrapped up.
Once I confirm it works on my multiplatform test project, I will release 5.0.0.RC2.

**Comment by sksamuel** [comment]
@charleskorn other than the JS node issue with the compiler, all looking good I think.

The multiplatform example project runs against latest snapshots no problem.
https://github.com/kotest/kotest-examples-multiplatform

**Comment by charleskorn** [comment]
Looks like everything is working now, thanks @sksamuel!

**Comment by sksamuel** [comment]
Awesome, 5.0.0.RC2 is out too.

Will wait on the JS reply from jetbrains before releasing 5.0.0 final

**Comment by sksamuel** [comment]
I have created a new issue to track the JS compiler bug, and am closing this.

https://github.com/kotest/kotest/issues/2683
